plugins/inputs/sqlserver/sqlserver.go

```diff
@@ -244,11 +247,36 @@ func (s *SQLServer) Gather(acc telegraf.Accumulator) error {
 }
 
 func (s *SQLServer) gatherServer(server string, query Query, acc telegraf.Accumulator) error {
-	// deferred opening
-	conn, err := sql.Open("mssql", server)
-	if err != nil {
-		return err
+	// deferred connection opening
+	var conn *sql.DB
+
+	if strings.Contains(strings.ToLower(server), strings.ToLower("password=")) {
+		// when password is provided in connection string, use SQL auth
+
+		var err error
+		conn, err = sql.Open("mssql", server)
+
+		if err != nil {
+			return err
+		}
+	} else {
+		// when password is not provided in connection string, use AAD auth with an MSI token
+
+		tokenProvider, err := getMSITokenProvider()
```

Comment:
Do we need to get the token for all the servers or can we get the token once and leverage it for all servers since t is from the same VM?